lisa/runners/lisa_runner.py

```diff
@@ -290,30 +290,46 @@ def _prepare_environments(self) -> None:
             try:
-                self.platform.deploy_environment(environment)
-                assert (
-                    environment.status == EnvironmentStatus.Deployed
-                ), f"actual: {environment.status}"
-                self._reset_awaitable_timer("deploy")
-            except ResourceAwaitableException as identifier:
-                if self._is_awaitable_timeout("deploy"):
-                    self._log.info(
-                        f"[{environment.name}] timeout on waiting for more resource: "
-                        f"{identifier}, skip assigning case."
+                try:
+                    # Attempt to deploy the environment
+                    self.platform.deploy_environment(environment)
+                    assert (
+                        environment.status == EnvironmentStatus.Deployed
+                    ), f"actual: {environment.status}"
+                    self._reset_awaitable_timer("deploy")
+                    break
+                except ResourceAwaitableException as identifier:
+                    if self._is_awaitable_timeout("deploy"):
+                        self._log.info(
+                            f"[{environment.name}] timeout on waiting for more "
+                            f"resource: {identifier}, skip assigning case."
+                        )
+                        raise SkippedException(identifier)
+                    else:
+                        # rerun prepare to calculate resource again.
+                        environment.status = EnvironmentStatus.New
+            except Exception as identifier:
+                if attempt >= max_retries - 1:
+                    # Final attempt failed; handle the failure
+                    self._attach_failed_environment_to_result(
+                        environment=environment,
+                        result=test_results[0],
+                        exception=identifier,
+                    )
+                    self._delete_environment_task(
+                        environment=environment, test_results=[]
                     )
-                    raise SkippedException(identifier)
                 else:
-                    # rerun prepare to calculate resource again.
-                    environment.status = EnvironmentStatus.New
-        except Exception as identifier:
-            self._attach_failed_environment_to_result(
-                environment=environment,
-                result=test_results[0],
-                exception=identifier,
-            )
-            self._delete_environment_task(environment=environment, test_results=[])
+                    # Log retry information and continue to the next attempt
+                    self._log.info(
+                        f"[{environment.name}] Deployment failed, retrying... "
+                        f"(Attempt {attempt + 1}/{max_retries - 1})"
+                    )
+                    environment.status = EnvironmentStatus.Prepared
```

Comment:
`environment.status = EnvironmentStatus.New` is also applicable here.

Reply:
with set it as new, I saw below exception, during retry deploy

```
2024-09-12 23:52:13.320[15700][ERROR] lisa.runner[0] case failed
Traceback (most recent call last):
  File "C:\code\lisa-lsg\lsg-lisa\lisa\lisa\runners\lisa_runner.py", line 298, in _deploy_environment_task
    self.platform.deploy_environment(environment)
  File "C:\code\lisa-lsg\lsg-lisa\lisa\lisa\platform_.py", line 184, in deploy_environment
    self._deploy_environment(environment, log)
  File "C:\code\lisa-lsg\lsg-lisa\lisa\lisa\sut_orchestrator\azure\platform_.py", line 607, in _deploy_environment
    raise identifier
  File "C:\code\lisa-lsg\lsg-lisa\lisa\lisa\sut_orchestrator\azure\platform_.py", line 578, in _deploy_environment
    location, deployment_parameters = self._create_deployment_parameters(
  File "C:\code\lisa-lsg\lsg-lisa\lisa\lisa\sut_orchestrator\azure\platform_.py", line 1118, in _create_deployment_parameters
    azure_node_runbook = self._create_node_runbook(
  File "C:\code\lisa-lsg\lsg-lisa\lisa\lisa\sut_orchestrator\azure\platform_.py", line 1276, in _create_node_runbook
    raise LisaException("vm_size is not detected before deploy")
lisa.util.LisaException: vm_size is not detected before deploy
```

Reply:
Can you check why it works for awaitable but not here?

Reply:
if encounter awaitable, I saw the prepare_environment is invoked again, so below two lines work when encounter deployment failure.
```
                    environment.status = EnvironmentStatus.New
                    self._prepare_environment(environment)
```

Reply:
It doesn't make sense, the logic in the same function. Why only Awaitable exception call prepare?

Reply:
for awaitable exception, it keep running, since is_done is always false, for deployment failure, is_done is true after running once.

![image](https://github.com/user-attachments/assets/5dccfe4c-c9b5-4a7c-96b0-0ea6815fc8f0)

Reply:
for Awaitable exception, is_done is always false, and it will keep running fetch_task, for retry logic, it only happens inside of _deploy_environment_task

Reply:
updated